sycl/doc/Assert.md

```diff
@@ -0,0 +1,250 @@
+
+int main() {
+  queue Q(ErrorHandler);
+  q.submit([&] (handler& CGH) {
+    CGH.parallel_for<class TheKernel>(range<2>{N, M}, [=](item<2> It) {
+      do_smth();
+      user_func(It);
+      do_smth_else();
+    });
+  });
+  Q.wait_and_throw();
+  std::cout << “One shouldn’t see this message.“;
+  return 0;
+}
+```
+
+In this use-case every work-item with even X dimension will trigger assertion
+failure. Assertion failure should be reported via asynchronous exceptions. If
+asynchronous exception handler is set the failure is reported with
+`cl::sycl::event_error` exception. Otherwise, SYCL Runtime should trigger abort.
+At least one failed assertion should be reported. The assertion failure message
+is printed to `stderr` by SYCL Runtime.
+
+When multiple kernels are enqueued and both fail at assertion at least single
+assertion should be reported.
+
+
+## User requirements
+
+From user's point of view there are the following requirements:
+
+| # | Title | Description | Importance |
+| - | ----- | ----------- | ---------- |
+| 1 | Handle assertion failure | Signal about assertion failure via SYCL asynchronous exception | Must have |
+| 2 | Print assert message | Assert function should print message to stderr at host | Must have |
+| 3 | Stop under debugger | When debugger is attached, break at assertion point | Highly desired |
+| 4 | Reliability | Assert failure should be reported regardless of kernel deadlock | Highly desired |
+
+
+## Contents of `cl::sycl::event_error`
+
+Interface of `cl::sycl::event_error` should look like:
+```
+class event_error : public runtime_error {
+public:
+  event_error() = default;
+
+  event_error(const char *Msg, cl_int Err)
+      : event_error(string_class(Msg), Err) {}
+
+  event_error(const string_class &Msg, cl_int Err) : runtime_error(Msg, Err) {}
+};
+```
+
+Regardless of whether asynchronous exception handler is set or not, there's an
+action to be performed by SYCL Runtime. To achieve this, information about
+assert failure should be propagated from device-side to SYCL Runtime. This
+should be performed via calls to `clGetEventInfo` for OpenCL backend and
+`zeEventQueryStatus` for Level-Zero backend.
+
+
+## Terms
+
+ - Device-side Runtime - part of device-code, which is supplied by Device-side
+   Compiler.
+ - Device-side Compiler - compiler which generates device-native binary image
+   based on input SPIR-V image.
+ - Low-level Runtime - the backend/runtime behind DPCPP Runtime.
+ - Accessor metadata - parts of accessor representation at device-side: pointer,
+   ranges, offset.
+
+
+## How it works?
+
+For the time being, `assert(expr)` macro ends up in call to
+`__devicelib_assert_fail` function. This function is part of [Device library extension](extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst#cl_intel_devicelib_cassert).
+Device code already contains call to the function. Currently, a device-binary
+is always linked against fallback implementation.
+
+
+### Device-specific approach
```

Comment:
I think the key differentiator between the "device specific" and "device agnostic" approaches is really the reliability. So I suggest to call them accordingly. E.g. "safe", "main", "primary" vs "fallback", "secondary".

Reply:
Employed safe-fallback wording.